WIP - feat/extended-download-dicomweb

TFRadicalImaging opened the pull request:
### 🚀 Add `download` Command: Full DICOM Study Downloader via DICOMweb

This PR introduces a new CLI command `download` that downloads an entire DICOM study—including study, series, instances, frames, and bulkdata—from a DICOMweb-compliant service.

The feature supports advanced functionality such as intelligent folder structuring, multi-frame handling, and JSON compression by default.

---

## 🚀 Features

- 📥 Download a **full DICOM study** including:
  - Study metadata
  - Series metadata
  - DICOM instances
  - Frames (multi-frame objects)
  - Bulk data (e.g., waveforms, PDFs, encapsulated binaries)
- 💾 Save everything locally in a clean, directory-based structure (SDW: *Static DICOM Web*)
- 🔍 Verbose output to show real-time progress (ideal for long studies)
- 🔌 Pluggable architecture for multiple DICOM sources (currently DICOMweb only)

---

## 🧪 Usage

```bash
node cli.js download <url> --StudyInstanceUID <UID> [options]
```

```bash
node cli.js download https://d14fa38qiwhyfd.cloudfront.net/dicomweb \
  --StudyInstanceUID 1.2.276.0.7230010.3.1.2.2155604110.4180.1021041295.21 \
  --directory ~/dcmjs
```

---

## 🛠️ Options

| Option | Description |
|--------|-------------|
| `<url>` | DICOMweb endpoint (must start with `http`) |
| `-S, --StudyInstanceUID` | The StudyInstanceUID you want to download |
| `-d, --directory` | Target folder for saving (default: `./`) |
| `--zip-json` | Save JSON metadata as compressed `.json.gz` |
| `--debug` | Enable debug logging (more verbose output) |

---

## 🗂️ Directory Structure (SDW Format)

```
downloads/
└── <StudyInstanceUID>/
    ├── index.json                      # Study-level metadata
    ├── series/
    │   ├── index.json                  # List of series
    │   └── <SeriesInstanceUID>/
    │       ├── metadata                # Series-level metadata
    │       └── instances/
    │           ├── index.json          # List of instances
    │           └── <SOPInstanceUID>/
    │               ├── index.json      # Instance metadata
    │               ├── frames/         # Frame pixel data
    │               └── metadata/       # Optional frame metadata
    └── bulkdata/
        └── <hashed-folder>/
            └── <hash>.json             # Base64-encoded binary content
```

---

## 🧱 Architecture

- `cliDownload.js` — CLI command registration and main workflow
- `DicomAccess.js` — Abstract class defining a common interface
- `DicomWebAccess.js` — Implements fetching data from DICOMweb
- `DicomWebSeriesAccess.js` / `DicomWebStudyAccess.js` — Handle specific study/series queries
- `DicomStoreSDW.js` — Stores metadata and binary data in a structured format
- `DicomStoreAccess.js` — Abstract base for pluggable storage backends

---

## 📅 Future Plans

- Support for:
  - DIMSE C-GET/C-MOVE
  - Local DICOMDIR parsing
- Upload (STOW-RS) and query (QIDO-RS) features
- UI integration or Electron wrapper

---

wayfarer3130 committed:
feat: Add DICOMweb download

wayfarer3130 committed:
Add lint settings

wayfarer3130 committed:
fix: Add logging

wayfarer3130 commented:
I moved the original dcmjs-commands to dcmjs-org and renamed this one as dcmjs-commands-fork so that the base ownership of the repository is with dcmjs-org.

wayfarer3130 committed:
Merge remote-tracking branch 'origin/main' into feat/download-dicomweb

TFRadicalImaging committed:
WIP - Extended download dicomweb command

wayfarer3130 commented on bin/dicomwebjs.js:
```diff
@@ -1,35 +1,115 @@
+  })
+
+program
+  .command("download-v2")
```
Please just remove the v1 instance from this and call it download

wayfarer3130 commented on bin/dicomwebjs.js:
```diff
@@ -1,35 +1,115 @@
+    }
+  })
+
+program
```
Can you extract the individual program arguments to individual files?  Something like bin/cliDownload   I think you did that in another PR for me, and I liked how you did that there.

wayfarer3130 commented on bin/dicomwebjs.js:
```diff
@@ -1,35 +1,115 @@
+    let downloadDir = baseDir;
+    let counter = 1;
+
+    while (fs.existsSync(downloadDir)) {
```
This should not initiate a new download directory instance - the directory should be a dicomweb root directory, and the path should be based on the dicomweb standard path.

wayfarer3130 replied:
Also, the handling of the path setting should be standardized so that multiple commands can use the same logic for the directory.

TFRadicalImaging replied:
I'm not sure if I understand what you want here.
The result is already as you expect, doesn't?
I just added a check to not replace the study that already exists with the new one. So that way, it creates the same folder name just adding a "_1" / "_2" / "_3" to not replace the original one (only if already exists)

TFRadicalImaging replied:
Here is an example of output:
<img width="424" alt="Captura de Tela 2025-04-04 às 11 45 44" src="https://github.com/user-attachments/assets/598c85a9-7a05-4819-b248-77741764f8d6" />

wayfarer3130 commented on bin/dicomwebjs.js:
```diff
@@ -1,35 +1,115 @@
+
+    console.log(`📁 Study will be downloaded to: ${downloadDir}`);
+  
+    const dicomWebConfig = {
```
You can have a single function that adds/parses out the dicomWebConfig from the arguments, and a second one that adds/parses the output directory functions as config/output objects.

wayfarer3130 commented on bin/dicomwebjs.js:
```diff
@@ -1,35 +1,115 @@
+  
+    console.log(`🎉 Download complete. All accessible instances and metadata were saved.`);
+    console.log(`📁 Study available at: ${downloadDir}`);
+    console.log("💡 Some files may have been skipped due to server restrictions or unsupported SOP Classes.");
```
This should be indicated as a warn level issue when there is a problem.

wayfarer3130 commented on yarn.lock:
We should just use the bun lock so there is only one lock file.

TFRadicalImaging replied:
This yarn wasn't added by me. Also it's not using bun anywhere in the code. At least I couldn't find.

wayfarer3130 commented:
For this change, what I want is a generic interface for reading and writing DICOM data, so that it can read and write to any/all:
DICOMweb - using the dicomweb-client dcmjs library
DIMSE - using the dicom-dimse library
File - static dicomweb format - using the fs library
File - DICOMDir format - using the dcmjs library and the fs library for reading/writing

You don't have to IMPLEMENT all of these, the only ones required right now are reading from DICOMweb and store to SDW file format, but it should go through a class interface to set things up.    I would suggest testing with the SDW file for reading as well as writing, but the eventual goal is to have a single implementation that handles various endpoints.

wayfarer3130 commented:
export class DicomAccess {
  public static createInstance(url, options) {
    if (url.startsWith("scp:")) {
      return new ScpDicomAccess(url, options);
    }
    if (url.startsWith("http")) {
      return new DicomWebAccess(url, options);
    }
    return new FileDicomAccess(url, options);
  }

  public queryStudy?(studyInstanceUID: string): Promise<DicomStudyAccess>;
  public storeStudy(studyAccess: DicomStudyAccess, options): Promise<string>;
}

export class DicomStudyAccess {
  public abstract getPatientData();
  public abstract getStudyData();

  // Not quite sure yet what these will all return, as it should work for dimse
  // dicomweb, local dicomweb file system, local dicom part 10 directories (DICOMDIR?)
  public iterateSeries(options): Iterable<DicomSeriesAccess>;

  public storeSeries(seriesAccess: SeriesAccess) {
    // Call the series level store metadata
  }
}

export abstract class DicomSeriesAccess {
  public getMetadata?(): object[];

  /** Stores the data in the preferred format for the given output location */
  public storeInstance?(imageAccess, options);
}

// Implementation for DICOMweb (separate file)
// Should use library dicomweb-client to query back end data

export class DicomWebAccess extends DicomAccess {
  constructor(dicomWebUrl) {
    // stuff here to create a dicomweb
  }

  // Implementation of DicomAccess in terms of the library dicomweb-client
}

// Dimse implementation (TODO later, but think about this one as to how to make it work consistently.)
export class DimseAccess extends DicomAccess {
  // Implementation in terms of dicom-dimse library
}

// File implementation
export class FileDicomAccess extends DicomAccess {
  // Implementation in terms of local file system - should use async returns

  public async queryStudy(studyInstanceUID: string) {
    // Read the file index.json data and return a FileDicomStudyAccess instance
  }
}

export class FileStudyAccess extends DicomStudyAccess {
  // Implement storeMetadata
}

TFRadicalImaging committed:
feat: implement modular DICOMweb downloader CLI with pluggable architecture and SDW output

- Supports full DICOM study download via DICOMweb (WADO-RS)
- Downloads study metadata, series metadata, instances, frames, and bulk data
- Saves to local disk using SDW structure (Simple DICOM Web)
- Implements pluggable DicomAccess interface (currently DICOMweb only)
- Provides clear console logs for user feedback
- Includes robust file organization and metadata serialization
- Ready for future support of DIMSE and DICOMDIR backends

wayfarer3130 commented on src/access/DicomAccess.js:
```diff
@@ -0,0 +1,22 @@
+// Abstract base class for DICOM access implementations
+export class DicomAccess {
+  static async createInstance(url, options) {
+    if (url.startsWith("http")) {
```
You need to have at least two implementations - local dicomweb file system and dicomweb http as this should work for both types.

wayfarer3130 commented on src/access/DicomWebSeriesAccess.js:
```diff
@@ -0,0 +1,140 @@
+import * as dicomweb from "../dicomweb.js";
+import { createDicomWebConfig, fixBulkDataURI } from "../utils/index.js";
+
+const nonFrameSOPs = {
```
Please extract a method to test if an instance is a non image instance.

wayfarer3130 commented on src/store/DicomStoreSDW.js:
```diff
@@ -0,0 +1,144 @@
+import { saveJson } from "../utils/index.js";
+import { DicomStoreAccess } from "./DicomStoreAccess.js";
+
+export class DicomStoreSDW extends DicomStoreAccess {
```
The store should be a full, first class DicomAccess that can both read and write the file system, so I think you will have to make the additional methods public.

wayfarer3130 replied:
I mean public on the DicomAccess, but unimplemented by default.

wayfarer3130 commented on src/utils/getDicomFilesFromDirectory.js:
```diff
@@ -0,0 +1,9 @@
+import fs from "fs";
+import path from "path";
+
+export function getDicomFilesFromDirectory(directory) {
```
You should use one of the directory tree scanners to get files in that directory and recursively, and unless the glob pattern is *.dcm, it should match everything - there might be an exclusion pattern, to exclude jpg, txt etc.

wayfarer3130 commented on bin/cliDownload.js:
```diff
@@ -0,0 +1,75 @@
+
+  try {
+    // Create access instance (currently supports only DICOMweb)
+    console.log("🔌 Creating DICOM access instance...");
```
The download action should basically look like:

const access = await DicomAccess.createInstance(url, options);
const destination = await DicomAccess.createInstance(baseDir, options)

const studyAccess = await access.queryStudy()
const studyDest = await access.queryStudy({ create: true })

studyDest.writeStudy(studyAccess);

Then, internally studyAccess should have things iterated from/over that studyDest needs in order to write the study source.

That way, studyDest being a DICOMDIR containing part 10 files can be written to, as can a response to a DICOMweb part 10 query, or a DICOMweb query can all be written to independently of the source type.

wayfarer3130 commented on src/access/DicomWebSeriesAccess.js:
```diff
@@ -0,0 +1,140 @@
+    this.metadata = metadata;
+
+    // Containers to store queried data
+    this.seriesInstanceUIDsMetadata = new Map();
```
Don't assume that everything gets read up front, instead read things as they are requested, grouping them into maps of objects as required with an async function doing the await on each item as it arrives.

wayfarer3130 committed:
Update the interfaces for downloading

wayfarer3130 committed:
fix: Run dicomwebjs command and store as iterative store data

wayfarer3130 committed:
Partial refactoring to work as tree structured data

wayfarer3130 committed:
fix: Add generic child type

wayfarer3130 committed:
fix: Store metadata as downloaded.

wayfarer3130 committed:
fix: Update instance store to have frame and bulkdata store

wayfarer3130 committed:
Add bulkdata and instances writing

wayfarer3130 committed:
fix: Adding start of DICOMweb json retrieve

wayfarer3130 committed:
Use jsdom in favor of the XML http request

wayfarer3130 committed:
fix: Download studies from local SDW - mostly working

wayfarer3130 committed:
Fix issue with series query results

wayfarer3130 committed:
fix: Most of study download now working

wayfarer3130 committed:
Working download from both dicomweb and local static dicomweb to bulkdata

wayfarer3130 committed:
fix: Bulkdata download

wayfarer3130 commented:
Closing in favour of the original version of htis so that Thomas can review.
